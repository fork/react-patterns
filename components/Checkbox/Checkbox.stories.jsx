import React, { useState } from 'react';

import Checkbox from './Checkbox';

export default {
  title: 'Checkbox',
  component: Checkbox
};

export const Checked = () => <Checkbox label="Checkbox 1" id="1" value="true" />;
export const Unchecked = () => <Checkbox label="Checkbox 2" id="2" />;
export const Error = () => (
  <Checkbox label="Checkbox 3" id="3" required error errorMessage="Dieses Feld ist verpflichtend" />
);
export const Success = () => <Checkbox label="Checkbox 3" id="3" required value="true" />;
export const Disabled = () => <Checkbox label="Checkbox 4" id="4" disabled />;
export const WithoutLabel = () => <Checkbox label="Checkbox 5" id="5" hideLabel />;
export const withState = () => {
  const [checked, setChecked] = useState(false);
  return (
    <Checkbox
      label="Checkbox 6"
      id="6"
      required
      value={checked}
      onChange={() => setChecked(prev => !prev)}
    />
  );
};
