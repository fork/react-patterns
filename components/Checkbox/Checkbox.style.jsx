import styled, { css } from 'styled-components';

import { color, space, hideVisually } from '../../stylesheets';

const StyledCheckbox = styled.div`
  display: block;
  width: 100%;

  label {
    display: flex;
  }

  .checkbox__label {
    position: relative;
    display: flex;
    align-items: center;
    color: ${color('black')};
    cursor: pointer;
    hyphens: auto;
    padding-left: ${space('s')};

    ${({ hideLabel }) =>
      hideLabel &&
      css`
        ${hideVisually()}
      `}

    a {
      text-decoration: underline;
    }
  }

  .checkbox__custom-checkbox {
    cursor: pointer;
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    border: 1px solid ${color('black')};
    width: 20px;
    height: 20px;

    ${props =>
      props.error &&
      css`
        border-color: ${color('highlight')};
      `}

    ${props =>
      props.success &&
      css`
        border-color: ${color('secondary')}; ;
      `}

    svg {
      display: none;
      fill: #000;
      margin-left: 2px;

      ${props =>
        props.success &&
        css`
          fill: ${color('secondary')}; ;
        `}
    }
  }

  input {
    ${hideVisually()}

    &[disabled] ~ .checkbox__label, &[disabled] ~ .checkbox__custom-checkbox {
      opacity: 0.5;
      cursor: not-allowed;
    }

    &:checked ~ .checkbox__custom-checkbox svg {
      display: block;
    }
  }

  .checkbox__required {
    position: absolute;
    top: -10px;
    right: -10px;
    line-height: 1;
    width: 6px;
    height: 6px;
    color: ${color('black')};

    ${props =>
      props.error &&
      css`
        color: ${color('highlight')};
      `}

    ${props =>
      props.success &&
      css`
        color: ${color('secondary')}; ;
      `}
  }

  .checkbox__error {
    display: block;
    color: ${color('highlight')};
  }
`;

export default StyledCheckbox;
