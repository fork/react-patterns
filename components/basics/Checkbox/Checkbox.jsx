import PropTypes from 'prop-types';
import React from 'react';

import StyledCheckbox from './Checkbox.style';
import Icon from '../Icon';

const Checkbox = ({
  label,
  id,
  required,
  onChange,
  error,
  errorMessage,
  disabled,
  value,
  hideLabel,
  ...props
}) => (
  <StyledCheckbox error={error} success={required && value} hideLabel={hideLabel}>
    <label htmlFor={id}>
      <input
        id={id}
        name={id}
        type="checkbox"
        onChange={onChange}
        aria-invalid={error}
        disabled={disabled}
        checked={value}
        {...props}
      />
      <span className="checkbox__custom-checkbox">
        <Icon icon="check" />
        {required && <span className="checkbox__required">*</span>}
      </span>
      <span className="checkbox__label">{label}</span>
    </label>
    {error && errorMessage && <span className="checkbox__error">{errorMessage}</span>}
  </StyledCheckbox>
);

Checkbox.propTypes = {
  label: PropTypes.string,
  id: PropTypes.string.isRequired,
  required: PropTypes.bool,
  onChange: PropTypes.func,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
  disabled: PropTypes.bool,
  hideLabel: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
};

Checkbox.defaultProps = {
  label: null,
  required: false,
  onChange: null,
  error: false,
  errorMessage: null,
  disabled: false,
  value: false,
  hideLabel: false
};

export default Checkbox;
