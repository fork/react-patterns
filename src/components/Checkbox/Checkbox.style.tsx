import styled, { css } from 'styled-components';

import { color, space, hideVisually } from '../../stylesheets';

const StyledCheckbox = styled.div<{ error?: boolean; success?: boolean; hideLabel?: boolean }>`
  display: block;
  width: 100%;

  label {
    display: flex;
  }

  .checkbox__label {
    position: relative;
    display: flex;
    align-items: center;
    color: ${color('neutral')};
    cursor: pointer;
    hyphens: auto;
    padding-left: ${space('s')};

    ${({ hideLabel }) =>
      hideLabel &&
      css`
        ${hideVisually()}
      `}

    a {
      text-decoration: underline;
    }
  }

  .checkbox__custom-checkbox {
    cursor: pointer;
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    border: 1px solid ${color('neutral')};
    width: 20px;
    height: 20px;

    ${props =>
      props.error &&
      css`
        border-color: ${color('warning')};
      `}

    ${props =>
      props.success &&
      css`
        border-color: ${color('success')};
      `}

    svg {
      display: none;
      fill: #000;
      margin-left: 2px;

      ${props =>
        props.success &&
        css`
          fill: ${color('success')};
        `}
    }
  }

  input {
    ${hideVisually()}

    &[disabled] ~ .checkbox__label, &[disabled] ~ .checkbox__custom-checkbox {
      opacity: 0.5;
      cursor: not-allowed;
    }

    &:checked ~ .checkbox__custom-checkbox svg {
      display: block;
    }
  }

  .checkbox__required {
    position: absolute;
    top: -10px;
    right: -10px;
    line-height: 1;
    width: 6px;
    height: 6px;
    color: ${color('neutral')};

    ${props =>
      props.error &&
      css`
        color: ${color('warning')};
      `}

    ${props =>
      props.success &&
      css`
        color: ${color('success')};
      `}
  }

  .checkbox__error {
    display: block;
    color: ${color('warning')};
  }
`;

export default StyledCheckbox;
