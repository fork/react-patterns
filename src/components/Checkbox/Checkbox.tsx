import React from 'react';

import StyledCheckbox from './Checkbox.style';
import Icon from '../Icon';

export type CheckboxProps = {
  label: string;
  id: string;
  required?: boolean;
  onChange?: (ev: React.ChangeEvent<HTMLInputElement>) => void;
  error?: boolean;
  errorMessage?: string;
  disabled?: boolean;
  checked?: boolean;
  hideLabel?: boolean;
};

const Checkbox = ({
  label,
  id,
  required,
  onChange,
  error,
  errorMessage,
  disabled,
  checked,
  hideLabel
}: CheckboxProps) => (
  <StyledCheckbox error={error} success={required && checked} hideLabel={hideLabel}>
    <label htmlFor={id}>
      <input
        id={id}
        name={id}
        type="checkbox"
        onChange={onChange}
        aria-invalid={error}
        disabled={disabled}
        checked={checked}
      />

      <span className="checkbox__custom-checkbox">
        <Icon icon="check" />

        {required && <span className="checkbox__required">*</span>}
      </span>
      <span className="checkbox__label">{label}</span>
    </label>

    {error && errorMessage && <span className="checkbox__error">{errorMessage}</span>}
  </StyledCheckbox>
);

export default Checkbox;
