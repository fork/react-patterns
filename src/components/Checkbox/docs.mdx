import { Story, Canvas } from '@storybook/addon-docs';

# Checkbox

Eine Checkbox ist eine Schaltfläche, mit der Nutzer:innen zwischen zwei gegensätzlichen Zuständen oder Werten auswählen können. Checkboxen können eigenständig oder als Liste eingesetzt werden.

Mehrere Checkboxen in einer Liste schließen sich nicht gegenseitig aus: Nutzer:innen können keine, eine oder alle Optionen einer Liste wählen.

Checkboxen **werden verwendet**:

- in Formularen
- als Filter
- um von Nutzer:innen ein Einverständnis für Nutzungsbedingungen zu erhalten
- eine Option zu aktivieren oder zu deaktivieren

**Nicht eingesetzt** werden Checkboxen, wenn Nutzer:innen aus einer Liste nur eine Option auswählen können. Stattdessen sollten [Radio Buttons](?path=/docs/components-input-radio) oder ein [Select](?path=/docs/components-input-select) verwendet werden.

<Canvas>
  <Story id="components-input-checkbox--default" />
</Canvas>

<Canvas>
  <Story id="components-input-checkbox--disabled" />
</Canvas>

<Canvas>
  <Story id="components-input-checkbox--required" />
</Canvas>

<Canvas>
  <Story id="components-input-checkbox--success" />
</Canvas>

<Canvas>
  <Story id="components-input-checkbox--error" />
</Canvas>

## Anatomie

Eine Checkbox besteht aus dem eigentlichen Eingabefeld und einem Label. Optional kann eine Gruppe mehrerer Checkboxes ein Label erhalten.

<!-- Optional: schematische Darstellung -->

1. **Group label** (optional): Beschreibt den Inhalt der Gruppe
2. **Input**: Eingabefeld der Checkbox, zeigt den Zustand (_unselected_, _selected_, _indeterminate_) an
3. **Checkbox Label**: Beschreibt, was der:die Nutzer:in auswählt

<!--
## Platzierung

Hier kann beschrieben werden, wie eine Checkbox in verschiedenen Kontexten platziert wird.
-->

<!--
## Inhalt

Hier können Regeln für die Inhalte von Checkboxes definiert werden, z.B. Wording von Labels.
-->
