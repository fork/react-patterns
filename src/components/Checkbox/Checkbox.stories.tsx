import React, { useState } from 'react';
import { Story, Meta } from '@storybook/react';

import Checkbox, { CheckboxProps } from './Checkbox';
import docs from './docs.mdx';

export default {
  title: 'Components/Input/Checkbox',
  component: Checkbox,
  parameters: {
    docs: {
      page: docs
    }
  }
} as Meta;

const defaultData = {
  label: 'Checkbox a',
  id: 'a',
  required: false,
  checked: false,
  error: false,
  errorMessage: 'Dieses Feld ist verpflichtend',
  hideLabel: false
};

const Template: Story<CheckboxProps> = args => <Checkbox {...args} />;

const TemplateWithState: Story<CheckboxProps> = args => {
  const [checked, setChecked] = useState(false);
  return <Checkbox {...args} checked={checked} onChange={() => setChecked(prev => !prev)} />;
};

export const Checked = Template.bind({});
Checked.args = { ...defaultData, label: 'Checkbox 1', id: '1', checked: true };

export const Unchecked = Template.bind({});
Unchecked.args = { ...defaultData, label: 'Checkbox 2', id: '2', checked: false };

export const Error = Template.bind({});
Error.args = {
  label: 'Checkbox 3',
  id: '3',
  required: true,
  error: true,
  errorMessage: 'Dieses Feld ist verpflichtend'
};

export const Success = Template.bind({});
Success.args = { ...defaultData, label: 'Checkbox 4', id: '4', checked: true, required: true };

export const Disabled = Template.bind({});
Disabled.args = { ...defaultData, label: 'Checkbox 5', id: '5', disabled: true };

export const WithoutLabel = Template.bind({});
WithoutLabel.args = { ...defaultData, label: 'Checkbox 6', id: '6', hideLabel: true };

export const WithState = TemplateWithState.bind({});
WithState.args = { label: 'Checkbox 7', id: '7', required: true };
