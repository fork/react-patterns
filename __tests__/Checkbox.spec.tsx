import React from 'react';
import { shallow } from 'enzyme';

import Checkbox from '../src/components/Checkbox';

describe('Checkbox', () => {
  test('Checkbox should render correctly', () => {
    const component = shallow(<Checkbox label="Checkbox Label" id="1" />);

    expect(component).toMatchSnapshot();
  });

  test('Checkbox without label should render correctly', () => {
    const component = shallow(<Checkbox hideLabel id="1" label="Checkbox Label" />);

    expect(component).toMatchSnapshot();
  });

  test('Checkbox with default value should render correctly', () => {
    const component = shallow(<Checkbox hideLabel checked id="1" label="Checkbox Label" />);

    expect(component).toMatchSnapshot();
  });
});
