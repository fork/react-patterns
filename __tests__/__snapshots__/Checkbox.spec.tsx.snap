// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Checkbox should render correctly 1`] = `
<styled.div>
  <label
    htmlFor="1"
  >
    <input
      id="1"
      name="1"
      type="checkbox"
    />
    <span
      className="checkbox__custom-checkbox"
    >
      <Icon
        icon="check"
      />
    </span>
    <span
      className="checkbox__label"
    >
      Checkbox Label
    </span>
  </label>
</styled.div>
`;

exports[`Checkbox Checkbox with default value should render correctly 1`] = `
<styled.div
  hideLabel={true}
>
  <label
    htmlFor="1"
  >
    <input
      checked={true}
      id="1"
      name="1"
      type="checkbox"
    />
    <span
      className="checkbox__custom-checkbox"
    >
      <Icon
        icon="check"
      />
    </span>
    <span
      className="checkbox__label"
    >
      Checkbox Label
    </span>
  </label>
</styled.div>
`;

exports[`Checkbox Checkbox without label should render correctly 1`] = `
<styled.div
  hideLabel={true}
>
  <label
    htmlFor="1"
  >
    <input
      id="1"
      name="1"
      type="checkbox"
    />
    <span
      className="checkbox__custom-checkbox"
    >
      <Icon
        icon="check"
      />
    </span>
    <span
      className="checkbox__label"
    >
      Checkbox Label
    </span>
  </label>
</styled.div>
`;
